### PR TITLE
ENH: Add MNE-C style SNR estimate

### PR DESCRIPTION
**Pull request by Eric89GXL**
Moves us toward something like #1453.

### CODE REVIEW

File: mne/minimum_norm/inverse.py [hunk 1]
```diff
@@ -1439,3 +1440,73 @@ def compute_rank_inverse(inv):
...
+    bad = (snr <= 1)
+    lambda2_est[bad] = 100.
+    remaining[bad] = False
+
+    # parameters
+    lambda_mult = 0.9
+    sing2 = (inv['sing'] * inv['sing'])[:, np.newaxis]
+    val = chi2.isf(1e-3, n_ch - 1)
+    n_iter = 0
+    while n_iter < 1000 and remaining.any():
```
Comment (dengemann) [inline mne/minimum_norm/inverse.py]: hardcoded as well in MNE-C?

File: mne/viz/evoked.py [hunk 2]
```diff
@@ -527,3 +527,35 @@ def whitened_gfp(x, rank=None):
...
+    ax.plot(evoked.times, snr, color=[1, 0, 0])
+    ax.set_xlim(lims[:2])
+    ax.set_ylim(lims[2:])
+    ax.set_ylabel('SNR')
+    ax.set_xlabel('Time (sec)')
+    if evoked.comment is not None:
+        ax.set_title(evoked.comment)
+    plt.draw()
+    if show:
+        plt.show()
```
Comment (dengemann) [inline mne/viz/evoked.py]: should return a fig object

File: examples/inverse/plot_snr_estimate.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+print(__doc__)
+
+data_dir = op.join(data_path(), 'MEG', 'sample')
+fname_inv = op.join(data_dir, 'sample_audvis-meg-oct-6-meg-inv.fif')
+fname_evoked = op.join(data_dir, 'sample_audvis-ave.fif')
+
+inv = read_inverse_operator(fname_inv)
+evoked = read_evokeds(fname_evoked)[0]
+evoked.data -= evoked.data[:, evoked.times <= 0].mean(1)[:, np.newaxis]
```
Comment (dengemann) [inline examples/inverse/plot_snr_estimate.py]: what's the purpose of this line? why not use our baseling functions? ` mne.baseline.rescale`

**Comment by dengemann** [comment]
would be good to figure out how this is conceptually / pracrtically different from displaying the whitened global field power, see covariance examples.

File: mne/minimum_norm/inverse.py [hunk 4]
```diff
@@ -1439,3 +1440,73 @@ def compute_rank_inverse(inv):
...
+    data_white_ef = np.dot(inv['eigen_fields']['data'], data_white)
+    n_ch, n_times = data_white.shape
+
+    # Adapted from mne_analyze/regularization.c, compute_regularization
+    n_zero = (inv['noise_cov']['eig'] <= 0).sum()
+    logger.info('Effective nchan = %d - %d = %d'
+                % (n_ch, n_zero, n_ch - n_zero))
+    signal = np.sum(data_white ** 2, axis=0)  # sum of squares across channels
+    noise = n_ch - n_zero
+    snr = signal / noise
```
Comment (dengemann) [inline mne/minimum_norm/inverse.py]: indeed similar to the whitened global field power we have in evooed.plot_white

File: mne/minimum_norm/inverse.py [hunk 5]
```diff
@@ -1439,3 +1440,73 @@ def compute_rank_inverse(inv):
...
+        f[inv['sing'] == 0] = 0
+        ew = data_white_ef[:, remaining] * (1.0 - f)
+        # check condition
+        err = np.sum(ew * ew, axis=0)
+        remaining[np.where(remaining)[0][err < val]] = False
+        lambda2_est[remaining] *= lambda_mult
+        n_iter += 1
+    snr_est = 1.0 / np.sqrt(lambda2_est)
+    snr = np.sqrt(snr)
+    return snr, snr_est
```
Comment (dengemann) [inline mne/minimum_norm/inverse.py]: here we see the additional processing + outputs. We should think about how to communicate differences + integrate the two viz functions educationally.

File: mne/minimum_norm/inverse.py (hunk 1 above)
Comment (Eric89GXL) [inline mne/minimum_norm/inverse.py]: It is, and in my testing on one dataset it stopped at about 60 and on another at about 120, so if you get to 1000 there is  probably something strange going on.

File: mne/minimum_norm/inverse.py (hunk 5 above)
Comment (Eric89GXL) [inline mne/minimum_norm/inverse.py]: I agree, but I don't 100% understand it myself, outside of making sure it agrees with the C code results.

File: examples/inverse/plot_snr_estimate.py (hunk 3 above)
Comment (Eric89GXL) [inline examples/inverse/plot_snr_estimate.py]: Just didn't know about it, I'll switch it

Comment (dengemann) [inline examples/inverse/plot_snr_estimate.py]: ;-)

File: mne/minimum_norm/inverse.py (hunk 5 above)
Comment (dengemann) [inline mne/minimum_norm/inverse.py]: I bet the first output looks more or less exactly as the whitened global field power ... We should think a bit about this.

Comment (Eric89GXL) [inline mne/minimum_norm/inverse.py]: yes, I think it should based on the calculations

Comment (dengemann) [inline mne/minimum_norm/inverse.py]: what to do with that redundancy? integrate it somehow?

Comment (Eric89GXL) [inline mne/minimum_norm/inverse.py]: Good question... not sure.

Comment (dengemann) [inline mne/minimum_norm/inverse.py]: Let's discuss this with @agramfort 
Maybe we might want to split the sensor from the source SNR or might want to have one function that does both on inputs or come up with a good narrative that makes this redundancy indispensible.

File: examples/inverse/plot_snr_estimate.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+from mne.viz import plot_snr_estimate
+
+print(__doc__)
+
+data_dir = op.join(data_path(), 'MEG', 'sample')
+fname_inv = op.join(data_dir, 'sample_audvis-meg-oct-6-meg-inv.fif')
+fname_evoked = op.join(data_dir, 'sample_audvis-ave.fif')
+
+inv = read_inverse_operator(fname_inv)
+evoked = read_evokeds(fname_evoked)[0]
```
Comment (agramfort) [inline examples/inverse/plot_snr_estimate.py]: just use baseline param in read_evokeds

File: mne/minimum_norm/inverse.py [hunk 7]
```diff
@@ -1439,3 +1440,73 @@ def compute_rank_inverse(inv):
...
+    evoked : instance of Evoked
+        Evoked instance.
+    inv : instance of InverseOperator
+        The inverse operator.
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see mne.verbose).
+
+    Returns
+    -------
+    snr : ndarray, shape(n_times,)
```
Comment (agramfort) [inline mne/minimum_norm/inverse.py]: insert space after shape

File: mne/minimum_norm/inverse.py (hunk 1 above)
Comment (agramfort) [inline mne/minimum_norm/inverse.py]: I prefer the pattern

for n_iter in range(1000):
      ...
     break
else:
     warn.('Did not converge')

File: mne/minimum_norm/tests/test_snr.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
+# -*- coding: utf-8 -*-
```
Comment (agramfort) [inline mne/minimum_norm/tests/test_snr.py]: don't forget author + license

File: mne/minimum_norm/tests/test_snr.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+fname_evoked = op.join(s_path, 'sample_audvis-ave.fif')
+
+
+@testing.requires_testing_data
+@requires_mne
+def test_snr():
+    """Test SNR calculation"""
+    tempdir = _TempDir()
+    inv = read_inverse_operator(fname_inv)
+    evoked = read_evokeds(fname_evoked)[0]
```
Comment (agramfort) [inline mne/minimum_norm/tests/test_snr.py]: use baseline param

File: mne/viz/evoked.py (hunk 2 above)
Comment (agramfort) [inline mne/viz/evoked.py]: +1

File: mne/viz/tests/test_misc.py [hunk 10]
```diff
@@ -112,4 +115,14 @@ def test_plot_source_spectrogram():
                   [[1, 2], [3, 4]], tmax=7)
 
 
+@testing.requires_testing_data
+def test_plot_snr():
+    """Test plotting SNR estimate
+    """
+    inv = read_inverse_operator(inv_fname)
+    evoked = read_evokeds(evoked_fname)[0]
```
Comment (agramfort) [inline mne/viz/tests/test_misc.py]: baseline param

**Comment by agramfort** [comment]
besides LGTM

File: mne/minimum_norm/inverse.py (hunk 5 above)
Comment (dengemann) [inline mne/minimum_norm/inverse.py]: @agramfort what's your opinion / feeling here regarding redundancy to plot_white?

Comment (agramfort) [inline mne/minimum_norm/inverse.py]: I'd like to see a plot to see if it matches before scratching my head

File: mne/minimum_norm/tests/test_snr.py (hunk 8 above)
Comment (Eric89GXL) [inline mne/minimum_norm/tests/test_snr.py]: Sure, I'll add it. I don't think many of our `test_` files have this information, BTW...

File: mne/minimum_norm/inverse.py (hunk 5 above)
Comment (Eric89GXL) [inline mne/minimum_norm/inverse.py]: It looks pretty similar, the GFP plots done in `master` split by channel type, the one in this PR collapses. Should I refactor the two bits of code to use the same calculations? Then the test against C that I wrote won't work, so I'm not sure what I'd test against. (I'm not sure of any way to get the green curve out of the C code, I was hoping mne_compute_mne would do it, but the `--snronly` param just dumps the red GFP curve.)

Comment (agramfort) [inline mne/minimum_norm/inverse.py]: can you share some images?

**Comment by Eric89GXL** [comment]
Sample images using:
```
from os import path as op
import mne
import matplotlib.pyplot as plt
plt.ion()

data_dir = op.join(mne.datasets.sample.data_path(), 'MEG', 'sample')
fname_inv = op.join(data_dir, 'sample_audvis-meg-oct-6-meg-inv.fif')
fname_evoked = op.join(data_dir, 'sample_audvis-ave.fif')
fname_cov = op.join(data_dir, 'sample_audvis-cov.fif')

inv = mne.minimum_norm.read_inverse_operator(fname_inv)
evoked = mne.read_evokeds(fname_evoked, baseline=(None, 0))[0]
mne.viz.plot_snr_estimate(evoked, inv)
evoked.plot_white(mne.read_cov(fname_cov))
```
figure_1

figure_2

**Comment by agramfort** [comment]
a clear difference to understand is that SNR needs inv and plot white needs just the noise cov.

**Comment by dengemann** [comment]
> a clear difference to understand is that SNR needs inv and plot white needs just the noise cov.

but if I'm not mistaken the first output of the snr plot is only based on the parts of the inv which come from the cov, essentially the same computations we would do for SSS. 

We need some SNR tutorial that explains the different levels of diagnostic plotting. The plot_white is a calibration plot for the whitening and allows to detect violations, etc., but also gives an SNR estimate. The SNR plot ... does what in addition? We need to make this explicit.

**Comment by agramfort** [comment]
it's not clear to me

I did not take the time to look

**Comment by Eric89GXL** [comment]
@dengemann yes, that is true of the first output (SNR) but not of the second, which requires the eigen fields and singular values from decomposition. And now that I think about it, we probably *need* to use `inv['whitener']` and not one computed a different way (as in `whiten_evoked`) in order to ensure the calculations actually make sense, so I'm -1 on trying to unify those bits of code.

**Comment by agramfort** [comment]
can we see the maths of plot_snr somewhere?

**Comment by Eric89GXL** [comment]
yeah `estimate_snr` in `inverse.py`

**Comment by Eric89GXL** [comment]
https://github.com/mne-tools/mne-python/pull/1844/files#diff-629cc1d6799fbae9ecd4fd5f53b84a43R1449

**Comment by agramfort** [comment]
good numpy == latex :)

**Comment by Eric89GXL** [comment]
Oh good, I was going to check that out, too. Where is it in the manual? Maybe I can add a bit of docstring, or feel free to make a PR into my branch if you want to

**Comment by agramfort** [comment]
no idea / no time :( :(

**Comment by Eric89GXL** [comment]
Ohh, I thought you meant that you checked and said "good" because "numpy == latex" (code == manual), I didn't realize you meant that the numpy was going to be readable like Latex. I get it now...

I'll see if I can find the defs in the manual.

**Comment by Eric89GXL** [comment]
Okay I figured out what's going on and added it to the `Notes` section of the docstring. Feel free to fire up your favorite introspecting editor and look at the description, see if it makes sense to you.

**Comment by Eric89GXL** [comment]
This is ready for review/merge from my end @dengemann @agramfort.

File: mne/viz/evoked.py [hunk 11]
```diff
@@ -527,3 +527,36 @@ def whitened_gfp(x, rank=None):
...
+    """Plot a data SNR estimate
+
+    Parameters
+    ----------
+    evoked : instance of Evoked
+        The evoked instance. This should probably be baseline-corrected.
+    inv : instance of InverseOperator
+        The minimum-norm inverse operator.
+    show : bool
+        Whether to show the figure or not. Defaults to True.
```
Comment (agramfort) [inline mne/viz/evoked.py]: Returns is missing

**Comment by agramfort** [comment]
besides +1 for merge.

let's wait for you to get the second +1

**Comment by Eric89GXL** [comment]
@agramfort done, @dengemann feel free to merge if you're also happy

**Comment by dengemann** [comment]
We should try to avoid red green plots as default, mind the color blind people. Why not red and blue? We also need a more practical description of how to interpret the estimated SNR? What does it practically mean. We can take care later of the second issue, but then we should open an issue right now.

**Comment by Eric89GXL** [comment]
I'm fine changing that -- I used red/green to mimic MNE-C, but don't need to.

**Comment by Eric89GXL** [comment]
Colors fixed. Feel free to merge and open an issue @dengemann

**Comment by dengemann** [comment]
Thanks a lot Eric!

**Pull request merged**